src/Microsoft.ML.Data/Transforms/ValueMappingTransformer.cs

```diff
@@ -0,0 +1,995 @@
+                    {
+                        TValue value = values.ElementAt(i);
+                        if(!keyTypeValueMapping.ContainsKey(value))
+                        {
+                            keyTypeValueMapping.Add(value, index);
+                            index++;
+                        }
+
+                        var keyValue = keyTypeValueMapping[value];
+                        indices[i] = keyValue;
+                    }
+
+                    dataViewBuilder.AddColumn(valueColumnName, GetKeyValueGetter(metaKeys), 0, metaKeys.Count(), indices);
+                }
+            }
+            else
+            {
+                dataViewBuilder.AddColumn(valueColumnName, valueType, values.ToArray());
+            }
+
+            return dataViewBuilder.GetDataView();
+        }
+    }
+
+    /// <summary>
+    /// The ValueMappingTransformer is a 1-1 mapping from a key to value. The key type and value type are specified
+    /// through TKey and TValue. Arrays are supported for vector types which can be used as either a key or a value
+    /// or both. The mapping is specified, not trained by providiing a list of keys and a list of values.
+    /// </summary>
+    /// <typeparam name="TKey">Specifies the key type</typeparam>
+    /// <typeparam name="TValue">Specifies the value type</typeparam>
+    public sealed class ValueMappingTransformer<TKey, TValue> : ValueMappingTransformer
+    {
+        /// <summary>
+        /// Constructs a ValueMappingTransformer with a key type to value type
+        /// </summary>
+        /// <param name="env">Instance of the host environment</param>
+        /// <param name="keys">The list of keys that are TKey</param>
+        /// <param name="values">The list of values that are TValue</param>
+        /// <param name="treatValuesAsKeyTypes">Specifies to treat the values as a <see cref="KeyType"/></param>
+        /// <param name="columns">The specified columns to apply</param>
+        public ValueMappingTransformer(IHostEnvironment env, IEnumerable<TKey> keys, IEnumerable<TValue> values, bool treatValuesAsKeyTypes, (string Input, string Output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingTransformer<TKey, TValue>)),
+                  ConvertToDataView(env, keys, values, treatValuesAsKeyTypes), KeyColumnName, ValueColumnName, columns)
+        { }
+
+        /// <summary>
+        /// Constructs a ValueMappingTransformer with a key type to value array type
+        /// </summary>
+        /// <param name="env">Instance of the host environment</param>
+        /// <param name="keys">The list of keys that are TKey</param>
+        /// <param name="values">The list of values that are TValue[]</param>
+        /// <param name="columns">The specified columns to apply</param>
+        public ValueMappingTransformer(IHostEnvironment env, IEnumerable<TKey> keys, IEnumerable<TValue[]> values, (string Input, string Output)[] columns)
```

Comment:
>I [](start = 123, length = 1)

why are those capitalized?

Reply:
Sorry it was originally Source and Name and then I converted to Input Output. I have changed all instances to input, output

---
In reply to: [243494506](https://github.com/dotnet/machinelearning/pull/1710#discussion_r243494506) [](ancestors = 243494506)